ethereum/ethevents/events.go

```diff
@@ -326,76 +333,66 @@ func (ev *EthereumEvents) processEventLogsFromTo(ctx context.Context,
 	return nil
 }
 
-func (ep *EventProcessor) id(e *ethtypes.Log) string {
-	return fmt.Sprintf("%x%d", e.TxHash, e.TxIndex)
+func (ep *EventProcessor) id(event *ethtypes.Log) string {
+	return fmt.Sprintf("%x%d", event.TxHash, event.TxIndex)
 }
 
-func (ep *EventProcessor) add(e *ethtypes.Log) {
-	eventID := ep.id(e)
-	ep.eventQueue.Store(eventID, logEvent{event: e, added: time.Now()})
+func (ep *EventProcessor) add(event *ethtypes.Log) {
+	eventID := ep.id(event)
+	ep.eventQueueLock.Lock()
+	defer ep.eventQueueLock.Unlock()
+	ep.eventQueue[eventID] = timedEvent{event: event, added: time.Now()}
 }
 
-func (ep *EventProcessor) del(e *ethtypes.Log) {
-	eventID := ep.id(e)
-	ep.eventQueue.Delete(eventID)
+func (ep *EventProcessor) del(event *ethtypes.Log) {
+	eventID := ep.id(event)
+	ep.eventQueueLock.Lock()
+	defer ep.eventQueueLock.Unlock()
+	delete(ep.eventQueue, eventID)
 }
 
 // next returns the first log event rady to be processed
 func (ep *EventProcessor) next() *ethtypes.Log {
-	nextId := ""
-	ep.eventQueue.Range(func(k, v interface{}) bool {
-		elog, ok := v.(logEvent)
-		if !ok {
-			log.Errorf("stored type is not %T", logEvent{})
-		}
-		if time.Since(elog.added) < ep.EventProcessThreshold {
-			return true
+	ep.eventQueueLock.Lock()
+	defer ep.eventQueueLock.Unlock()
+	for id, event := range ep.eventQueue {
+		if time.Since(event.added) >= ep.EventProcessThreshold {
+			delete(ep.eventQueue, id)
+			return event.event
 		}
-		nextId = k.(string)
-		return false
-	},
-	)
-	if nextId == "" {
-		return nil
 	}
-	next, ok := ep.eventQueue.LoadAndDelete(nextId)
-	if !ok {
-		log.Errorf("could not load and delete %s", nextId)
-		return nil
-	}
-	return next.(logEvent).event
+	return nil
 }
 
 func (ev *EthereumEvents) runEventProcessor(ctx context.Context) {
 	ev.EventProcessor.eventProcessorRunning = true
 	go func() {
-		var evt ethtypes.Log
-		var evtJSON []byte
-		var err error
 		for {
-			evt = <-ev.EventProcessor.Events
-			if evtJSON, err = evt.MarshalJSON(); err != nil {
+			event := <-ev.EventProcessor.Events
+			evtJSON, err := event.MarshalJSON()
+			if err != nil {
 				log.Error(err)
 				continue
 			}
-			if evt.Removed {
+			if event.Removed {
 				log.Warnf("removing reversed log event: %s", evtJSON)
-				evtc := evt
-				ev.EventProcessor.del(&evtc)
+				ev.EventProcessor.del(&event)
 			} else {
 				log.Debugf("queued event log: %s", evtJSON)
-				evtc := evt
-				ev.EventProcessor.add(&evtc)
+				ev.EventProcessor.add(&event)
 			}
 		}
 	}()
 
 	for {
+		// TODO(mvdan): Perhaps refactor this so we don't need a sleep.
+		// Maybe use a queue-like structure sorted by added time,
+		// and separately mark the removed events to ignore them.
 		time.Sleep(2 * time.Second)
```

Comment:
since this next is now safe we could reduce the sleep to something like `time.Milisecond*200`